### PR TITLE
Refactor indexing functions

### PR DESCRIPTION
**Pull request by t4c1**
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py src/test/unit`
- [ ] Run cpplint: `make cpplint`
- [ ] Declare copyright holder and open-source license: see below

#### Summary

This is a copy of #3010, just moved to original repo to make Jenkins formatting bot work.

Refactors indexing functions to simplify them including:
 - renames `lvalue.hpp` to `assign.hpp` and `lvalue_varmat.hpp` to `assign_varmat.hpp` to match functions implemented in them
 - deletes unused `rvalue_return` metaprogram
 - changes signatures of `rvalue` and `assign` functions:
 - deletes unused argument depth
 - removes default value from name argument to allow for the next change
 - replaces `cons_index_list` structure with (possibly variadic) argument list of indices. This makes `cons_index_list` and related functions redundant so they are removed.

#### Intended Effect

The generated code that uses indexing is more readable.

#### How to Verify

The changes are tested by changed tests.

#### Side Effects

This changes the interface of indexing functions. So it needs to be merged together with changes in stanc that will generate code according to the new interface.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Tadej Ciglarič

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Review by bbbales2 — APPROVED** [review]
Looks good. I left two notes but I think they were minor

File: src/stan/model/indexing/rvalue.hpp [hunk 1]
```diff
@@ -729,20 +647,19 @@ inline auto rvalue(
...
           require_std_vector_t<StdVec>* = nullptr>
-inline auto rvalue(StdVec&& v, const cons_index_list<index_uni, Idx>& idxs,
-                   const char* name = "ANON", int depth = 0) {
-  math::check_range("array[uni, ...] index", name, v.size(), idxs.head_.n_);
+inline auto rvalue(StdVec&& v, const char* name, index_uni idx1,
+                   const Idxes&... idxes) {
+  math::check_range("array[uni, ...] index", name, v.size(), idx1.n_);
   if (std::is_rvalue_reference<StdVec>::value) {
-    return rvalue(std::move(v[idxs.head_.n_ - 1]), idxs.tail_, name, depth + 1);
+    return rvalue(std::move(v[idx1.n_ - 1]), name, idxes...);
```
[the remark this answers is not in the document]
Comment (t4c1) [reply src/stan/model/indexing/rvalue.hpp:660]: `std::forwards` requires the type of the thing to forward to be specified. Determining appropriate value type with appropriate reference would be just as much code as this is. Something like
```
return rvalue(std::forward<std::conditional<std::is_rvalue_reference<StdVec>::value, value_type_t<StdVec>&&, value_type_t<StdVec>&>>(v[idx1.n_ - 1]), name, idxes...);
```

**Comment by SteveBronder** [comment]
I skimmed over the code and I think everything looks okie dokie! I'll have another look tonight

File: src/stan/model/indexing/assign.hpp [hunk 2]
```diff
@@ -889,22 +797,20 @@ inline void assign(T&& x, const nil_index_list& /* idxs */, U&& y,
...
- * @param[in] depth Indexing depth (default 0).
+ * @param[in] name Name of variable
+ * @param[in] idx1 uni index
+ * @param[in] idxes Remaining indices
  * @throw std::out_of_range If any of the indices are out of bounds.
  * @throw std::invalid_argument If the dimensions do not match in the
  * tail assignment.
  */
-template <typename StdVec, typename Idx, typename U,
+template <typename StdVec, typename U, typename... Idxes,
```
Comment (SteveBronder) [inline src/stan/model/indexing/assign.hpp:808]: Typo?
```suggestion
template <typename StdVec, typename U, typename... Idxs,
```

File: src/stan/model/indexing/assign.hpp [hunk 3]
```diff
@@ -945,29 +847,30 @@ inline void assign(StdVec&& x,
...
-template <typename T, typename Idx1, typename Idx2, typename U,
-          require_all_std_vector_t<T, U>* = nullptr>
-inline void assign(T&& x, const cons_index_list<Idx1, Idx2>& idxs, U&& y,
-                   const char* name = "ANON", int depth = 0) {
-  int x_idx_size = rvalue_index_size(idxs.head_, x.size());
+template <typename T, typename Idx1, typename... Idxes, typename U,
+          require_all_std_vector_t<T, U>* = nullptr,
+          require_not_same_t<Idx1, index_uni>* = nullptr>
+inline void assign(T&& x, U&& y, const char* name, const Idx1& idx1,
+                   const Idxes&... idxes) {
```
Comment (SteveBronder) [inline src/stan/model/indexing/assign.hpp:863]: same as above `idxs`

File: src/stan/model/indexing/assign_varmat.hpp [hunk 4]
```diff
@@ -84,34 +79,31 @@ inline void assign(VarVec&& x,
...
   Eigen::Matrix<double, -1, 1> y_vals(assign_size);
   std::unordered_set<int> x_set;
   x_set.reserve(assign_size);
   // We have to use two loops to avoid aliasing issues.
   for (int i = assign_size - 1; i >= 0; --i) {
-    if (likely(x_set.insert(idxs.head_.ns_[i]).second)) {
-      stan::math::check_range("vector[multi] assign", name, x_size,
-                              idxs.head_.ns_[i]);
-      x_idx[i] = idxs.head_.ns_[i] - 1;
+    if (likely(x_set.insert(idx.ns_[i]).second)) {
```
Comment (SteveBronder) [inline src/stan/model/indexing/assign_varmat.hpp:104]: I wonder if we should remove the likely here? I've been writing things lately that have a lot of repeated values so idk it's a judgement call

File: src/stan/model/indexing/assign_varmat.hpp [hunk 5]
```diff
@@ -200,53 +187,51 @@ template <typename Mat1, typename Vec,
...
+                              col_idx.ns_[i]);
+      x_idx[i] = col_idx.ns_[i] - 1;
+      prev_val.coeffRef(i) = x.val().coeffRef(row_idx_val, x_idx[i]);
       y_vals.coeffRef(i) = y.val().coeff(i);
     } else {
       x_idx[i] = -1;
     }
   }
   for (int i = assign_cols - 1; i >= 0; --i) {
     if (likely(x_idx[i] != -1)) {
```
Comment (SteveBronder) [inline src/stan/model/indexing/assign_varmat.hpp:216]: Same here for likely()

**Review by SteveBronder — APPROVED** [review]
Looks so much nicer :pray: couple comments but overall looks good!

File: src/stan/model/indexing/assign.hpp [hunk 6]
```diff
@@ -89,27 +83,24 @@ inline void assign(Vec&& x,
...
  * @throw std::invalid_argument If the value size isn't the same as
  * the indexed size.
  */
 template <typename Vec1, typename Vec2,
           require_all_eigen_vector_t<Vec1, Vec2>* = nullptr>
-inline void assign(Vec1&& x,
-                   const cons_index_list<index_multi, nil_index_list>& idxs,
-                   const Vec2& y, const char* name = "ANON", int depth = 0) {
+inline void assign(Vec1&& x, const Vec2& y, const char* name,
+                   const index_multi& idx) {
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply src/stan/model/indexing/assign.hpp:96]: Once you go variadic, default value isnt really that useful I guess, the only use is assign(x,y) without indexing. The name is always set in code-gen.

File: src/stan/model/indexing/assign.hpp [hunk 7]
```diff
@@ -222,19 +201,15 @@ inline void assign(Vec1&& x,
...
  * @throw std::invalid_argument If the value size isn't the same as
  * the indexed size.
  */
 template <typename Vec1, typename Vec2,
           require_all_vector_t<Vec1, Vec2>* = nullptr,
           require_all_not_std_vector_t<Vec1, Vec2>* = nullptr>
-inline void assign(Vec1&& x,
-                   const cons_index_list<index_omni, nil_index_list>& idxs,
-                   Vec2&& y, const char* name = "ANON", int depth = 0) {
+inline void assign(Vec1&& x, Vec2&& y, const char* name, index_omni) {
```
[the remark this answers is not in the document]
Comment (t4c1) [reply src/stan/model/indexing/assign.hpp:212]: done

File: src/stan/model/indexing/assign.hpp [hunk 8]
```diff
@@ -311,19 +280,15 @@ inline void assign(Mat1&& x,
...
  * @throw std::out_of_range If any of the indices are out of bounds.
  * @throw std::invalid_argument If the dimensions of the indexed
  * matrix and right-hand side matrix do not match.
  */
 template <typename Mat1, typename Mat2,
           require_all_dense_dynamic_t<Mat1, Mat2>* = nullptr>
-inline void assign(Mat1&& x,
-                   const cons_index_list<index_omni, nil_index_list>& idxs,
-                   Mat2&& y, const char* name = "ANON", int depth = 0) {
+inline void assign(Mat1&& x, Mat2&& y, const char* name, index_omni) {
```
[the remark this answers is not in the document]
Comment (t4c1) [reply src/stan/model/indexing/assign.hpp:291]: done

File: src/stan/model/indexing/assign.hpp (hunk 2 above)
Comment (t4c1) [reply src/stan/model/indexing/assign.hpp:808]: fixed

File: src/stan/model/indexing/assign.hpp (hunk 3 above)
Comment (t4c1) [reply src/stan/model/indexing/assign.hpp:863]: fixed

File: src/stan/model/indexing/assign_varmat.hpp (hunk 4 above)
Comment (t4c1) [reply src/stan/model/indexing/assign_varmat.hpp:104]: No idea.

**Comment by rok-cesnovar** [comment]
The test error popping up now is legit: https://jenkins.mc-stan.org/blue/organizations/jenkins/Stan/detail/PR-3011/12/pipeline

Seems like a missing signature?

**Pull request merged**